enhancements/cluster-scope-secret-volumes/shared-resources-via-openshift-builds-operator.md

```diff
@@ -0,0 +1,291 @@
+   kind: OpenShiftBuild
+   metadata:
+     name: cluster
+   spec:
+     shipwright:
+       build:
+         enabled: false
+     sharedResources:
+       enabled: false
+   ```
+
+3. The OpenShift Builds Operator reconciles the `OpenShiftBuild` instance (singleton), and reports
+   an appropriate status. If `spec.sharedResources.enabled` is `true`, the operator deploys the
+   Shared Resource CSI driver, custom resource definitions, and webhook.
+4. The operator will likewise deploy and manage Shipwright components using the `spec.shipwright.*`
+   fields.
+5. The operator will not enable any operand by default.
+
+
+### Variations
+
+- If OpenShift Pipelines is not installed and `spec.shipwright.build.enabled` is set to `true`,
+  the operator does not install Shipwright Builds and reports that OpenShift Pipelines should be
+  installed in its status.
+- If the operator is deployed on a cluster with the `CSIDriverSharedResource` feature gate enabled,
+  changing `spec.sharedResources.enabled` will not be respected, and an appropriate status
+  condition will report why the operator has not deployed the Shared Resource CSI driver.
+
+
+### API Extensions
+
+The OpenShift Builds will intoroduce a new configuration CRD, `OpenShiftBuild`. It will contain 
+`spec` fields that configure the Shared Resource CSI driver in addition to Shipwright-related
+components. The operator will also deploy (directly or indirectly) the following:
+
+- Custom resources for the Shipwright Operator
+- Custom resources and webhooks for Shipwright Build
+- Custom resources and webhooks for the Shared Resource CSI Driver
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+As noted above, the operator will not deploy the Shared Resource CSI Driver if the cluster has
+the respective OCP feature gate enabled. This generally applies to tech preview clusters or
+clusters with custom fature gate tunings.
+
+With the start of OCP 4.16, the Shared Resource CSI driver operator will be removed from the
+[Cluster Storage Operator](https://github.com/openshift/cluster-storage-operator). The feature
+will likewise be removed from OpenShift's feature sets - any code that needs the most up to date
+set of feature sets will need to be updated (ex - openshift/library-go).
+The Shared Resource CSI driver itself will need to be removed from the OpenShift payload. Its
+components will need to be re-productized as an OLM operand.
+
+Notably, the design of this operator does **_not_** intend to use OLM API dependency resolution
+to automate the deployment and management of OpenShift Pipelines. Cluster admimins will need to
+separately (and manually) install OpenShift Pipelines.
+
+
+### Risks and Mitigations
+
+#### Duplicate CSI Drivers
+
+On a tech preview cluster, the Same CSI driver could be deployed twice on tech preview clusters. The operator will check if the `CSIDriverSharedResource` feature is enabled before
+deploying the Shared Resource CSI driver itself.
```

Comment:
would it be better/safer to check for the presence of the CSI driver itself, rather than indirectly making the determination based on feature gates?

Reply:
also what happens if someone installs this operator/enables the csi driver and *then* turns on the techpreview featureset?

Reply:
can the new build operator "disable" the `CSIDriverSharedResource` feature gate that is included in the the tech preview one per https://github.com/openshift/api/blob/master/config/v1/types_feature.go#L169 *after* tech preview is turned on?

otherwise, I wonder it simply renaming the shared resource CSI driver to something different than what the tech preview installs would suffice here.  Then, any OCP builds that define CSI volumes would point to the new driver and that would bypass the tech perview version of the driver.

WDYT @adambkaplan @bparees ?

I'm also good with adding a CSI drvier instance check on top of the feature gate check.

WDYT @adambkaplan ?

Reply:
> can the new build operator "disable" the CSIDriverSharedResource feature gate that is included in the the tech preview one per https://github.com/openshift/api/blob/master/config/v1/types_feature.go#L169 after tech preview is turned on?

i don't think so, the a featureset (such as "techpreviewnoupgrade") is basically a static set of featuregates.  you can't turn off a specific gate, or remove it from the featureset (without shipping a new ocp release)

> otherwise, I wonder it simply renaming the shared resource CSI driver to something different than what the tech preview installs would suffice here. 

that could work, i don't know what impact is has on the buildconfig integration work that was done though.  might need to be some new/rework done there if it was renamed.

Reply:
ACK on the feature gate.

On the integration, the *corev1.CSIVolumeSource type is used directly in the build type per https://github.com/openshift/api/blob/master/build/v1/types.go#L1457C6-L1457C29 so it is up the user to specify the driver name in the BuildConfig.  There are no code changes in the OCP/OCM build controller to switch driver names.

Reply:
ah, ok, so it was generic support for CSI volumes that was added, not something specific to shared resource volumes.  cool

Reply:
I like using a different name for registering the CSI Driver, for two reasons:

1. Avoiding the driver name and deployment collision with OCP.
2. Branding - the current driver name has the `openshift.io` domain/api group.  Using a different api group (ex: `sharedresources.dev`) will make the core idea more upstream-friendly.